Change `value` to be `uintnat`

dra27 opened the pull request:
Historically, value has been treated as a signed integer which might be a pointer. This commit changes it to be an unsigned integer and thus a pointer which might be a signed integer.

This has the benefit that pointer arithmetic assertions always make sense, even if the actual address space spans 0x80000000 or 0x8000000000000000 and only requires annotations in the rarer instances
where value is actually a signed integer. This fixes several instances of invalid pointer assertion code in the debug runtime.

At present, this only comes into play when `CAML_INTERNALS` has been defined. User code may opt for the change by ensuring that `CAML_UNSIGNED_VALUE` has been defined before `caml/mlvalues.h` is included.

/cc @mshinwell, @stedolan

dra27 committed:
Change value to be uintnat

Historically, value has been treated as a signed integer which might be
a pointer. This commit changes it to be an unsigned integer and thus a
pointer which might be a signed integer.

This has the benefit that pointer arithmetic assertions always make
sense, even if the actual address space spans 0x80000000 or
0x8000000000000000 and only requires annotations in the rarer instances
where value is actually a signed integer.

This is clearly a breaking change and so only comes into play when
CAML_INTERNALS has been defined. User code may opt for the change by
ensuring that CAML_UNSIGNED_VALUE has been defined before mlvalues.h is
included.

gasche commented:
Is there an executive summary for non-C-wizards of what this means in terms of Undefined Behavior in the implementation? (Do arithmetic overflows becomes more or less undefined, for example?)

xavierleroy commented on byterun/caml/mlvalues.h:
```diff
@@ -71,12 +79,20 @@ typedef uintnat mark_t;
 /* Conversion macro names are always of the form  "to_from". */
 /* Example: Val_long as in "Val from long" or "Val of long". */
 #define Val_long(x)     ((intnat) (((uintnat)(x) << 1)) + 1)
```
In the `CAML_UNSIGNED_VALUE` case there should be no cast to `intnat`.

xavierleroy commented:
You're missing a bunch of casts in the bytecode interpreter, for operations such as division, modulus, right shifts, and integer comparisons.

xavierleroy commented:
Also the motivations should be clearer.  The first reason that I see for treating values as unsigned integers is to ensure that arithmetic overflows are defined as wrapping around instead of causing undefined behavior and weird optimizations.

I don't understand the issues with "pointer assertions" that you mention; could you give an example?

dra27 commented:
@gasche - my understanding, also as a non-C-wizard, is that the idea is that the semantics (or lack of them) are unchanged - everywhere where signed overflow may occur the `value` should already have been cast back from `uintnat` to `intnat`.

@xavierleroy - I am aware, but do not fully understand, that there are higher philosophical reasons for wanting `value` to be unsigned but my motivation was that we broke 32-bit builds (with #1070) because we were not testing them and when I turned on testing for it I got a bunch of assertion failures which I tracked down to pointers with bit 31 set used in the context `CAMLassert (pointer-above-0x80000000 > pointer-below-0x80000000)`. To me, it seemed easier to hone in on the more limited number of cases where a `value` is in fact a signed integer than locate and review all the memory layout assertions in the runtime. An example of this was line 464 of byterun/freelist.c where `cur` was negative and `bp` was positive.

I'm puzzled (and, believe me, I start from the assumption that it's me!), what casts are missing in byterun/interp.c? They need to be signed integers, and they all come from `Long_val` which is `(intnat)`?

xavierleroy commented:
Concerning byterun/interp.c, I remembered the code as less clean (cast-wise) than it actually is!  Not all integer arithmetic goes through `Long_val / Val_long`, see e.g. `ADDINT`, but explicit casts between `value` and `intnat` are there to make the intent clear.

So, I take back my comment that there is a problem in byterun/interp.c.  But I also take back my other comment that this change (of `value` to `uintnat`) removes undefined behaviors in the bytecode interpreter.  Taking `ADDINT` as an example, the undefined behavior on overflow is there because we ask for a signed addition via a cast to `intnat`.  If we cast to `uintnat` instead, we get unsigned addition and defined behavior on overflow.  That's something to keep in mind, but it would be a completely local change, independent from the proposed PR.

xavierleroy commented:
Concerning the assert failures, do you have a list of them, by chance?  My gut feeling is that the assertions are missing some `Bp_val` coercions, so that they compare things that are morally pointers at an integer type instead of at a pointer type.

xavierleroy commented:
Concerning undefined behavior in general, this PR introduces a major undefined behavior, namely defining functions with `uintnat` parameters and results, then claiming to the outside world that they have `intnat` parameter and results. I don't know of any contemporary ABI where this could break, but I can imagine some.

DemiMarie commented:
@gasche Unsigned overflow is defined (as wrapping) by the C standard.  Signed overflow is undefined behavior, but OCaml requires that it too wrap (and passes flags such as `-fwrapv` to the C compiler to ensure that it does).

xavierleroy commented:
I have cold feet about this proposal, for reasons mentioned above, so I move to close it.  However, I'm interested in some hints as to which GC assertions fail because `value` is a signed type, so that  or others can try to improve the assertions.

damiendoligez commented:
I got a private mail with details from @dra27 so I'll try to debug those assertions and report here.

dra27 committed:
Fix caml_fl_cur_wsz accounting

In caml_fl_add_blocks, caml_fl_cur_wsz was only updated with the size of
the first block, which causes problems on real 32-bit systems and some
hypothetical 64-bit systems with a lot of RAM...

caml_fl_add_blocks now walks the list and so fully updates caml_fl_cur_wsz.

dra27 committed:
Re-enable x86 debug runtime testing

dra27 commented:
Sorry for the slowness getting back to this one.

@xavierleroy - the culprits are the assertions on lines 463 and 464 of byterun/freelist.c. I haven't checked to see if they're failing repeatably, but I'm seeing large numbers of tests failing when run with the debug runtime from a mixture of those two.

@damiendoligez pointed me in the direction of an actual bug, which is in `caml_fl_add_blocks` - previously, it only added the size of the first block, not the entire chain of blocks. I've put together a tentative fix. It seems to work, but the world should be terrified that I just altered the garbage collector in any way...

dra27 commented:
@damiendoligez - presumably it would be much better rather than walking the list of blocks in `caml_fl_add_blocks` if `expand_heap` instead wrote the total size of the list to Field 2 of the first block?

damiendoligez commented:
Here is the solution I'd come up with:
```
commit 8cc70bd5357e0ba482b0b84dc265d9bb5c019adc
Author: Damien Doligez <damien.doligez@inria.fr>
Date:   Thu Oct 5 17:07:11 2017 +0200

    fix freelist size computation on 32-bit architectures

diff --git a/byterun/freelist.c b/byterun/freelist.c
index 26c1d9c..74efbe8 100644
--- a/byterun/freelist.c
+++ b/byterun/freelist.c
@@ -531,12 +531,18 @@ header_t *caml_fl_merge_block (value bp)
    [bp] must point to a list of blocks chained by their field 0,
    terminated by Val_NULL, and field 1 of the first block must point to
    the last block.
+
+   The blocks must be chained in order of increasing addresses,
+   and they must be contiguous. Thus the total size will be the
+   last block's size plus the difference between the addresses of
+   the last and first.
 */
 void caml_fl_add_blocks (value bp)
 {
   CAMLassert (fl_last != Val_NULL);
   CAMLassert (Next (fl_last) == Val_NULL);
-  caml_fl_cur_wsz += Whsize_bp (bp);
+  caml_fl_cur_wsz += (value *) Field (bp, 1) - (value *) bp
+                     + Whsize_bp (Field (bp, 1));
 
   if (bp > fl_last){
     Next (fl_last) = bp;
```
But I like your solution better. Walking the list is not a big deal, as these are maximal-size blocks so there won't be many of them.

dra27 commented:
Rebased onto #1483

xavierleroy commented:
As discussed at the developers meeting last week, we will not change `value` to be `uintnat`.  The other two commits in this PR are in #1483.  So I'm closing this PR.
